Build is broken for target aarch64-linux-android

The bionic interface slightly differs from glibc and others. The build is broken due to some type conflict:

```
error[E0605]: non-primitive cast: `&[u8; 129]` as `&[i8]`
  --> src/core/deviceinfo.rs:44:35
   |
44 |         let uuid = str_from_c_str(&ioctl.uuid as &[i8]).ok_or_else(|| {
   |                                   ^^^^^^^^^^^^^^^^^^^^ an `as` expression can only be used to convert between primitive types or to coerce to a specific trait object

error[E0605]: non-primitive cast: `&[u8; 128]` as `&[i8]`
  --> src/core/deviceinfo.rs:71:32
   |
71 |                 str_from_c_str(&ioctl.name as &[i8])
   |                                ^^^^^^^^^^^^^^^^^^^^ an `as` expression can only be used to convert between primitive types or to coerce to a specific trait object

error[E0308]: mismatched types
  --> src/core/dm.rs:98:59
   |
98 |         let _ = name.as_bytes().read(mut_slice_from_c_str(&mut hdr.name))?;
   |                                                           ^^^^^^^^^^^^^ expected slice `[i8]`, found array `[u8; 128]`
   |
   = note: expected mutable reference `&mut [i8]`
              found mutable reference `&mut [u8; 128]`

error[E0308]: mismatched types
   --> src/core/dm.rs:103:59
    |
103 |         let _ = uuid.as_bytes().read(mut_slice_from_c_str(&mut hdr.uuid))?;
    |                                                           ^^^^^^^^^^^^^ expected slice `[i8]`, found array `[u8; 129]`
    |
    = note: expected mutable reference `&mut [i8]`
               found mutable reference `&mut [u8; 129]`

error[E0308]: mismatched types
   --> src/core/dm.rs:481:44
    |
481 |             let dst = mut_slice_from_c_str(&mut targ.target_type);
    |                                            ^^^^^^^^^^^^^^^^^^^^^ expected slice `[i8]`, found array `[u8; 16]`
    |
    = note: expected mutable reference `&mut [i8]`
               found mutable reference `&mut [u8; 16]`

error[E0308]: mismatched types
   --> src/core/dm.rs:586:50
    |
586 |                 let target_type = str_from_c_str(&targ.target_type)
    |                                                  ^^^^^^^^^^^^^^^^^ expected slice `[i8]`, found array `[u8; 16]`
    |
    = note: expected reference `&[i8]`
               found reference `&[u8; 16]`

error[E0308]: mismatched types
  --> src/core/dm_ioctl.rs:41:33
   |
41 |                 &str_from_c_str(&self.name as &[libc::c_char]).unwrap_or("Could not parse string"),
   |                                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected `i8`, found `u8`
   |
   = note: expected reference `&[i8]`
              found reference `&[u8]`

error[E0308]: mismatched types
  --> src/core/dm_ioctl.rs:45:33
   |
45 |                 &str_from_c_str(&self.uuid as &[libc::c_char]).unwrap_or("Could not parse string"),
   |                                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected `i8`, found `u8`
   |
   = note: expected reference `&[i8]`
              found reference `&[u8]`

error[E0308]: mismatched types
  --> src/core/dm_ioctl.rs:49:33
   |
49 |                 &str_from_c_str(&self.data as &[libc::c_char]).unwrap_or("Could not parse string"),
   |                                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected `i8`, found `u8`
   |
   = note: expected reference `&[i8]`
              found reference `&[u8]`

error: aborting due to 9 previous errors

Some errors have detailed explanations: E0308, E0605.
For more information about an error, try `rustc --explain E0308`.
error: could not compile `devicemapper`

To learn more, run the command again with --verbose.
```

To reproduce run a `cargo check --target aarch64-linux-android` I tested with NDK r19c.

--- a/.github/workflows/main.yml
+++ b/.github/workflows/main.yml
@@ -56,6 +56,30 @@ jobs:
       - name: Test ${{ matrix.task }} on ${{ matrix.toolchain }} toolchain
         run: ${{ matrix.task }}
 
+  android-build:
+    env:
+      TARGET: aarch64-linux-android
+    runs-on: ubuntu-18.04
+    container:
+      image: docker.io/circleci/android:api-30-ndk
+      options: --user 0:0
+    steps:
+      - uses: actions/checkout@v2
+      - name: Install dependencies
+        run: |
+          sudo apt-get -q update
+          sudo apt-get -y install libclang-dev
+      - uses: actions-rs/toolchain@v1
+        with:
+          components: cargo
+          toolchain: stable
+          target: x86_64-unknown-linux-gnu
+          override: true
+      - name: Add Android target
+        run: rustup target add ${{ env.TARGET }}
+      - name: Cross compilation build on Android
+        run: cargo build --target ${{ env.TARGET }}
+
   checks-with-ci-repo:
     strategy:
       matrix:

--- a/src/core/deviceinfo.rs
+++ b/src/core/deviceinfo.rs
@@ -4,6 +4,7 @@
 
 use std::convert::TryFrom;
 
+use libc::c_char;
 use semver::Version;
 
 use crate::{
@@ -41,7 +42,7 @@ impl TryFrom<dmi::Struct_dm_ioctl> for DeviceInfo {
     type Error = DmError;
 
     fn try_from(ioctl: dmi::Struct_dm_ioctl) -> DmResult<Self> {
-        let uuid = str_from_c_str(&ioctl.uuid as &[i8]).ok_or_else(|| {
+        let uuid = str_from_c_str(&ioctl.uuid as &[c_char]).ok_or_else(|| {
             DmError::Dm(
                 ErrorEnum::Invalid,
                 "Devicemapper UUID is not null terminated".to_string(),
@@ -68,7 +69,7 @@ impl TryFrom<dmi::Struct_dm_ioctl> for DeviceInfo {
             // encoding is only 32 bits.
             dev: Device::from_kdev_t(ioctl.dev as u32),
             name: DmNameBuf::new(
-                str_from_c_str(&ioctl.name as &[i8])
+                str_from_c_str(&ioctl.name as &[c_char])
                     .ok_or_else(|| {
                         DmError::Dm(
                             ErrorEnum::Invalid,

--- a/src/core/util.rs
+++ b/src/core/util.rs
@@ -4,6 +4,8 @@
 
 use std::{mem::size_of, slice, str};
 
+use libc::c_char;
+
 /// The smallest number divisible by `align_to` and at least `num`.
 /// Precondition: `align_to` is a power of 2.
 /// Precondition: `num` + `align_to` < usize::MAX + 1.
@@ -14,13 +16,13 @@ pub fn align_to(num: usize, align_to: usize) -> usize {
     (num + agn) & !agn
 }
 
-/// Convert from a &[i8] to a &[u8].
-pub fn byte_slice_from_c_str(c_str: &[i8]) -> &[u8] {
+/// Convert from a &[c_char] to a &[u8].
+pub fn byte_slice_from_c_str(c_str: &[c_char]) -> &[u8] {
     unsafe { slice::from_raw_parts(c_str as *const _ as *const u8, c_str.len()) }
 }
 
 /// Return a String parsed from the C string up to the first \0, or None
-pub fn str_from_c_str(slc: &[i8]) -> Option<&str> {
+pub fn str_from_c_str(slc: &[c_char]) -> Option<&str> {
     let slc = byte_slice_from_c_str(slc);
     str_from_byte_slice(slc)
 }
@@ -33,7 +35,7 @@ pub fn str_from_byte_slice(slc: &[u8]) -> Option<&str> {
 }
 
 /// Return a mutable slice from the mutable C string provided as input
-pub fn mut_slice_from_c_str(c_str: &mut [i8]) -> &mut [u8] {
+pub fn mut_slice_from_c_str(c_str: &mut [c_char]) -> &mut [u8] {
     unsafe { slice::from_raw_parts_mut(c_str as *mut _ as *mut u8, c_str.len()) }
 }
 

